php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+
+```php
+use Temporal\Workflow;
+
+class UploadWorkflow extends Workflow\Workflow
+{
+    public function run(string $input): string
+    {
+        $result = yield $this->activities
+            ->withStartToFinish(100)
+            ->activityName($input)
+            ->get(Obj::class);
+
+        return $result->getValue();
+    }
+}
+```
+
+> No common interface required.
+
+### Registration
+Similar to activities each workflow type must be registered within worker context, since worklow might expose more that 
+one method (query, primary method) meta declation is required. 
+
+```php
+$worker = new WorkflowWorker();
+
+$workflow = (new WorkflowDeclaration('name'))
+                ->setClass(WofkflowClass::class)
+                ->setHandlerMethod('run')
+                ->addQueryHandler('name', 'method-name')
+                ->addSignalHandler('name', 'method-name');
+
+$worker->addWorkflow($workflow);
+```
+
+### Using Annotations
+In more advanced frameworks it should be possible to register workflows using annotations or attributes.
+
+```php
+use Temporal\Workflow\Annotation as Temporal;
+
+/**
+ * @Temporal\Workflow(name="name", handler="run")
+ */
+class Workflow 
+{
+    public function run(ContextInterface $ctx, string $input): string
+    {
+        // ...   
+    }
+
+    /**
+     * @Workflow\QueryMethod(name="something")
+     */
+    public function querySomething(string $world): string 
+    {
+        // ...
+    }
+    
+    /**
+     * @Workflow\SignalMethod(name="someSignal")
+     */
+    public function handleSignal(string $input): string 
+    {
+        // ...
+    }
+}
+``` 
+
+> The actual annotation format is TBD as it irrelevant to the implementation and might vary from framework to framework. 
+
+### Syntax and Atomic Blocks
+The building blocks of workflow is `yield` commands which might return actual value or a promise for this value. Every
+command wrapped via dedicated class:
+
+```php
+function run()
+{
+    yield new Activity('name', 'values', [/* options */]);
+}
+```
+
+Some commands might automatically return sub-commands. For example blocking activity execution:
+
+```php
+$result = yield (new Activity('name', 'values', [/* options */]))->get();
+```
+
+To wait for multiple activities at the same time we can introduce higher level promises such as:
+
+```php
+$a = yield new Activity('a', 'values', [/* options */]);
+$b = yield new Activity('b', 'values', [/* options */]);
+
+yield new WaitAll($a, $b);
+
+// or 
+yield new WaitAny($a, $b);
+```
+
+> More atomic blocks can be added down the road.
+
+#### Syntax Sugar
+It should be possible to mock `new Activity` and other commands with shorted methods from abstract workflow implementation:
+
+```php
+class MyWorkflow extends Workflow 
+{
+    public function run()
+    {
+        $result = yield $this->activities->doSomething()->get();
+        yield $this->sleep(1);
```

Comment:
What about adding sleep to the workflow context or use similar approach as in activities:

`yield Workflow::sleep(ctx, 1);`

Reply:
Sure. 

Under the hood, each yield command should be handled by the local workflow coordinator so the actual syntax to generate such command does not matter:

```php
yield $this->sleep(1); // same
yield new Sleep(1); // same
yield Workflow::sleep(1); // same
```

We do not need to pass $ctx since the coordinator can only process one workflow command at once (there are no threads in PHP for that). This means that the coordinator always knows what workflow yield command and what is the context. So, essentially, it's always passed inexplicably.